### PR TITLE
Add tracing and JankStats

### PR DESCRIPTION
**Pull request by keyboardsurfer**
Enable performance monitoring with JankStats and improve inspection results with Trace sections for some parts of the app.

### CODE REVIEW

File: core-navigation/build.gradle.kts [hunk 1]
```diff
@@ -28,6 +28,7 @@ dependencies {
     api(libs.androidx.hilt.navigation.compose)
     api(libs.androidx.navigation.compose)
 
+    implementation(libs.androidx.tracing.ktx)
```
Comment (tunjid) [inline core-navigation/build.gradle.kts:31]: Files in `core-navigation` don't seem to use any imports from `tracing`, can it be removed?

File: app/build.gradle.kts [hunk 2]
```diff
@@ -24,6 +24,7 @@ plugins {
     id("jacoco")
     id("dagger.hilt.android.plugin")
     id("nowinandroid.spotless")
+    id("com.google.firebase.firebase-perf") version "1.4.1"
```
Comment (JolandaVerhoef) [inline app/build.gradle.kts:27]: Why are you adding an explicit version here? Can we extract that version as we do with other library/plugin versions?

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 3]
```diff
@@ -22,19 +22,44 @@ import androidx.activity.compose.setContent
...
+        super.onResume()
+        isTrackingEnabled(true)
+    }
+
+    override fun onPause() {
+        super.onPause()
+        isTrackingEnabled(false)
+    }
+
+    private fun isTrackingEnabled(enabled: Boolean) {
```
Comment (JolandaVerhoef) [inline app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:62]: To be honest, I think this method is a bit overkill. Can you just inline the two occurences?

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt [hunk 4]
```diff
@@ -22,19 +22,44 @@ import androidx.activity.compose.setContent
...
 import com.google.samples.apps.nowinandroid.ui.NiaApp
 import dagger.hilt.android.AndroidEntryPoint
+import javax.inject.Inject
 
 @OptIn(ExperimentalMaterial3WindowSizeClassApi::class)
 @AndroidEntryPoint
 class MainActivity : ComponentActivity() {
+
+    @Inject
+    lateinit var lazyStats: dagger.Lazy<JankStats>
```
Comment (JolandaVerhoef) [inline app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:35]: Can you add some link to documentation here? Then when people open this file and are wondering about it, they can easily learn more.

File: app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt [hunk 5]
```diff
@@ -65,6 +67,7 @@ import com.google.samples.apps.nowinandroid.navigation.TopLevelDestination
 fun NiaApp(windowSizeClass: WindowSizeClass) {
     NiaTheme {
         val navController = rememberNavController()
+
```
Comment (JolandaVerhoef) [inline app/src/main/java/com/google/samples/apps/nowinandroid/ui/NiaApp.kt:70]: nit; remove

File: feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt [hunk 6]
```diff
@@ -57,18 +57,35 @@ import com.google.samples.apps.nowinandroid.core.designsystem.icon.NiaIcons
...
+            if (isScrollInProgress) {
+                metricsHolder.state?.addState(
+                    "ForYou:AuthorsCarousel:Scrolling",
+                    "Index=$firstVisibleItemIndex"
+                )
+            } else {
+                metricsHolder.state?.removeState("ForYou:AuthorsCarousel:Scrolling")
+            }
+        }.collect()
+    }
```
Comment (JolandaVerhoef) [inline feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt:86]: I'm not a fan of having the tracing code so close to the UI part. It clutters the composable definitions. Can we somehow extract this so inside the composables themselves you simply call `traceLazyList(lazyListState)` and be done with it?

**Review by JolandaVerhoef — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt (hunk 3 above)
Comment (keyboardsurfer) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:62]: Done

File: app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt (hunk 4 above)
Comment (keyboardsurfer) [reply app/src/main/java/com/google/samples/apps/nowinandroid/MainActivity.kt:35]: Added.

File: app/build.gradle.kts (hunk 2 above)
Comment (keyboardsurfer) [reply app/build.gradle.kts:27]: Had never done that before. Added a new plugin for this.

File: feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt (hunk 6 above)
Comment (keyboardsurfer) [reply feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt]: Created TrackScrollJank to take care of this.

**Comment by JolandaVerhoef** [comment]
This branch has merge conflicts, can you resolve those?

**Comment by keyboardsurfer** [comment]
All resolved.

File: feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt [hunk 7]
```diff
@@ -58,17 +59,22 @@ import com.google.samples.apps.nowinandroid.core.designsystem.icon.NiaIcons
...
+import com.google.samples.apps.nowinandroid.core.ui.TrackScrollJank
 
 @Composable
 fun AuthorsCarousel(
     authors: List<FollowableAuthor>,
     onAuthorClick: (String, Boolean) -> Unit,
     modifier: Modifier = Modifier
 ) {
+    val lazyListState = rememberLazyListState()
+    TrackScrollJank(lazyListState, "ForYou:AuthorsCarousel")
```
Comment (JolandaVerhoef) [inline feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/AuthorsCarousel.kt:71]: About the naming here, for macrobenchmark you're using something like `forYou:Feed` with a non-capital `f` as a `testTag`. Can you keep these kinds of strings consistent throughout the project?

File: feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt [hunk 8]
```diff
@@ -316,7 +323,11 @@ private fun TopicSelection(
     onTopicCheckedChanged: (String, Boolean) -> Unit,
     modifier: Modifier = Modifier
 ) = trace("TopicSelection") {
+    val lazyGridState = rememberLazyGridState()
+    TrackScrollJank(scrollableState = lazyGridState, stateName = "ForYou:TopicSelection")
```
Comment (JolandaVerhoef) [inline feature-foryou/src/main/java/com/google/samples/apps/nowinandroid/feature/foryou/ForYouScreen.kt:327]: Please be consistent with named parameters vs unnamed parameters

**Review by JolandaVerhoef — COMMENTED** [review]
We're adding a lot of production code here that might be unfamiliar to developers. Can we consider adding a learning journey, pointing to all these different pieces of the puzzle, and then point to the learning journey from within code comments?

**Comment by keyboardsurfer** [comment]
I agree with creating a learning journey for the performance additions and have tracked it in #181.

**Review by JolandaVerhoef — APPROVED** [review]

**Review by dturner — CHANGES_REQUESTED** [review]
Please add some supporting documentation for this plugin/feature, either by adding to the existing README or (better) start a performance tracking learning journey. 

The docs should include: 

- The events which are being monitored
- How the events are currently recorded, and how they could be sent to a Firebase backend (see below)
- An example output, what the output means and potentially how to address jank. For example, when running the app on the emulator I see the following output: 

```
V/NiA Jank: FrameData(frameStartNanos=13037274863989, frameDurationUiNanos=286913934, frameDurationCpuNanos=293708517, frameOverrunNanos=277093643, isJank=true, states=[Navigation: for_you_route])
```

But, as a layman, I'm not sure what this is telling me, or what (if anything) I should do about it. 

I'd also like to see a `prod` configuration which switches the event recording from using `Log` to using a real backend, although this can be done in a follow up PR.

**Comment by keyboardsurfer** [comment]
Thanks for the review, @dturner. I have added your change requests to #181, which we'll address ASAP.
Is it okay to land this CL and promise to work on the learning journey next?

**Comment by keyboardsurfer** [comment]
That's what the issue is for. I've also added it to my internal task tracker.

**Pull request merged**